libslide/src/grammar.rs

```diff
@@ -365,6 +382,21 @@ pub struct UnaryExpr<E: Expression> {
     pub rhs: Rc<E>,
 }
 
+impl<E> UnaryExpr<E>
+where
+    E: Expression,
+{
+    pub fn negate<T>(expr: T) -> Self
+    where
+        T: Into<Rc<E>>,
+    {
+        Self {
+            op: UnaryOperator::SignNegative,
+            rhs: expr.into(),
```

Comment:
shouldn't negate not always be sign negative. if you negating a negative it should flip

Reply:
the contructor is just to negate the given expression. I don't think this method should be responsible for unfolding the underlying expression to figure out if it needs to be normalized.